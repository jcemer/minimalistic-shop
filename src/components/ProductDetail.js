import React, { PropTypes } from 'react'

import AddToCartButton from './AddToCartButton'

const ProductDetail = ({ _id, image, price, description, stock, addToCart }) => {
  return (
    <div className="ui unstackable items">
      <div className="item">
        <div className="image">
          <img src={image} alt={`Product ${_id}`} />
        </div>
        <div className="content">
          <div className="ui header">Product {_id}</div>
          <div className="meta">
            Price {price}
          </div>
          <div className="description">
            Description: {description}
          </div>
          <div className="extra">
            Remaining stock: {stock.remaining} item(s)
          </div>
          <div className="extra">
            <AddToCartButton _id={_id} stock={stock} addToCart={addToCart} />
          </div>
        </div>
      </div>
    </div>
  )
}

ProductDetail.propTypes = {
  _id: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  stock: PropTypes.shape({
    remaining: PropTypes.number.isRequired
  }).isRequired,
  addToCart: PropTypes.func.isRequired,
}

export default ProductDetail
