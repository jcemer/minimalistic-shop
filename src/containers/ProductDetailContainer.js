import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import _find from 'lodash/find'

import { addToCart } from '../actions/cart'
import ProductDetail from '../components/ProductDetail'

export const ProductDetailContainer = ({ isReady, item, addToCart }) => {
  if (!isReady) return null
  if (!item) return <span>Product not found</span>

  return (
    <div className="ui container">
      <div className="ui breadcrumb">
        <Link className="section" to='/'>Home</Link>
        <i className="right angle icon divider"></i>
        <div className="active section">Product</div>
      </div>
      <ProductDetail addToCart={addToCart} {...item} />
    </div>
  )
}

export const mapStateToProps = (
  { products: { isReady, items } },
  { match: { params: { _id } } }
) => {
  if (!isReady) return { isReady: false, item: undefined }
  return { isReady, item: _find(items, { _id }) }
}

export default connect(mapStateToProps, { addToCart })(ProductDetailContainer)
