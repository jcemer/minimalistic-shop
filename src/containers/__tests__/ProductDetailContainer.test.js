import React from 'react'
import { shallow, render } from 'enzyme'
import toJson from 'enzyme-to-json'
import { MemoryRouter } from 'react-router-dom'

import { ProductDetailContainer, mapStateToProps } from '../ProductDetailContainer'
import ProductDetail from '../../components/ProductDetail'

describe('ProductDetailContainer', () => {
  describe('<Component />', () => {
    const addToCart = () => {}

    it('should not render if it is not ready', () => {
      const wrapper = shallow(<ProductDetailContainer addToCart={addToCart} />)

      expect(wrapper.children().exists()).toBe(false)
    })

    it('should render "not found" without item', () => {
      const wrapper = shallow(<ProductDetailContainer addToCart={addToCart} isReady />)

      expect(wrapper.contains('Product not found')).toBe(true)
    })

    it('should render ProductDetail with item', () => {
      const item = { _id: '123', image: 'url', price: '$1', description: 'bla', stock: { remaining: 1 } }
      const wrapper = shallow(<ProductDetailContainer addToCart={addToCart} isReady item={item} />)

      expect(wrapper.contains(<ProductDetail addToCart={addToCart} {...item} />)).toBe(true)
    })

    it('should render the proper HTML of ProductDetail with item', () => {
      const item = { _id: '123', image: 'url', price: '$1', description: 'bla', stock: { remaining: 1 } }
      const wrapper = render(
        <MemoryRouter>
          <ProductDetailContainer addToCart={addToCart} isReady item={item} />
        </MemoryRouter>
      )

      expect(toJson(wrapper)).toMatchSnapshot()
    })
  })

  describe('mapStateToProps', () => {
    it('should returns no item if products is not ready', () => {
      const props = mapStateToProps(
        { products: { isReady: false } },
        { match: { params: { _id: 1 } } }
      )

      expect(props).toEqual({ isReady: false, item: undefined })
    })

    it('should returns matched item by "_id" if products is ready', () => {
      const item = { _id: 1 }
      const props = mapStateToProps(
        { products: { isReady: true, items: [{ _id: 0 }, item] } },
        { match: { params: { _id: 1 } } }
      )

      expect(props).toEqual({ isReady: true, item })
    })
  })
})
