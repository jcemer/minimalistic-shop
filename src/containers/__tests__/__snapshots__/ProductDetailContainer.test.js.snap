exports[`ProductDetailContainer <Component /> should render the proper HTML of ProductDetail with item 1`] = `
<div
  class="ui container">
  <div
    class="ui breadcrumb">
    <a
      class="section"
      href="/">
      Home
    </a>
    <i
      class="right angle icon divider" />
    <div
      class="active section">
      Product
    </div>
  </div>
  <div
    class="ui unstackable items">
    <div
      class="item">
      <div
        class="image">
        <img
          alt="Product 123"
          src="url" />
      </div>
      <div
        class="content">
        <div
          class="ui header">
          Product 123
        </div>
        <div
          class="meta">
          Price $1
        </div>
        <div
          class="description">
          Description: bla
        </div>
        <div
          class="extra">
          Remaining stock: 1 item(s)
        </div>
        <div
          class="extra">
          <button
            class="ui green button">
            Add to cart
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
